tools/normalize.c

```diff
@@ -0,0 +1,177 @@
+			normalized_sample = sample_ratio * UINT16_MAX / 2;
+		} else {
+			normalized_sample = sample_ratio * UINT32_MAX / 2;
+		}
+
+		// Write the normalized sample
+		add_normalized_sample(channel, (uint32_t) normalized_sample);
+
+		// Subtract from the rolling average
+		if (!fifo_read(&channel->samples, &old_sample)) {
+			fprintf(stderr, "Tried to read from empty FIFO\n");
+			return -ENOENT;
+		}
+		channel->rolling_average -= (float) old_sample / (float) average_width;
+	}
+	return 0;
+	
+}
+
+void init_channel(struct channel_data *channel, union bl_msg_data *msg) {
+	memset(channel, 0, sizeof(*channel));
+	channel->msg.type = msg->channel_conf.sample32 ? BL_MSG_SAMPLE_DATA32 : BL_MSG_SAMPLE_DATA16;
+	channel->msg.channel = msg->sample_data.channel;
+
+}
+
+int read_stream(FILE *stream, long average_width) {
+	union bl_msg_data msg; // message for reading into
+	struct channel_data channels[BL_ACQ__SRC_COUNT]; // One for every possible channel
+	long average_width_samples = 0; // Average width measured in samples
+	while (!bl_sig_killed && bl_msg_yaml_parse(stream, &msg)) {
+		if (msg.type == BL_MSG_CHANNEL_CONF) {
+			init_channel(channels + msg.channel_conf.channel, &msg);
+			bl_msg_yaml_print(stdout, &msg);
+		} else if (msg.type == BL_MSG_START) {
+			average_width_samples = average_width * msg.start.frequency / 1000; // Magical 1000 from being measured in milliseconds.
+
+			if (average_width_samples > FIFO_MAX) {
+				unsigned max_time = (unsigned) (1000 * (float) FIFO_MAX / (float) msg.start.frequency);
+				fprintf(stderr, "Unable to process that many samples, can handle averaging over up to %d milliseconds for samples at frequency %dHz\n", max_time, msg.start.frequency);
+				return -1;
+			}
+			bl_msg_yaml_print(stdout, &msg);
+		} else if (msg.type == BL_MSG_SAMPLE_DATA16 || msg.type == BL_MSG_SAMPLE_DATA32) {
+			struct channel_data *channel = channels + msg.sample_data.channel;
+
+			if (channel->msg.type != msg.type) {
+				fprintf(stderr, "Error: Sample data for channel %d has an unexpected type (expected %d, got %d)\n", msg.channel_conf.channel, channel->msg.type, msg.type);
+				return -1;
+			}
+
+			for (unsigned i = 0; i < msg.sample_data.count; i++) {
+				int ret;
+				if (msg.type == BL_MSG_SAMPLE_DATA16) {
+					ret = add_sample(average_width_samples, channel, msg.sample_data.data16[i]);
+				} else {
+					ret = add_sample(average_width_samples, channel, msg.sample_data.data32[i]);
+				}
+				if (ret < 0) {
+					return ret;
+				}
+			}
+		} else {
+			bl_msg_yaml_print(stdout, &msg);
+		}
+	}
+	return 0;
+}
+
+void usage(FILE* file, char *argv[]) {
+	fprintf(file, "Usage: %s [AVERAGE_WIDTH]\n", argv[0]);
+	fprintf(file, "  AVERAGE_WIDTH: The time (in ms) to average over\n");
+}
+
+int main(int argc, char *argv[]) {
+	long average_width = DEFAULT_AVERAGE_WIDTH;
+	/* Reads from stdin, writes to stdout */
+	/* The only argument is averaging width, which has a default */
+	if (argc > 2) {
```

Comment:
I tend to avoid magic numbers for positional arguments with an anonymous enum, scoped to `main()`, something like:

```c
enum {
	ARG_PROG_NAME,
	ARG_AVERAGE_WIDTH,
	ARG__COUNT,
};
```

And use those instead of the 0s 1s and 2s.